File: frontend/Frontend/lib/models/gamification_models.dart
import 'package:json_annotation/json_annotation.dart';

part 'gamification_models.g.dart';

/// User profile with gamification metadata
@JsonSerializable()
class UserProfile {
  final String id;
  final String name;
  final String email;
  final DateTime joinedAt;
  final int dailyGoalMinutes;
  final int dailyGoalAyahs;
  final int level;
  final int xp;
  final int longestStreak;
  final String? avatarUrl;
  final int timezoneOffset; // Minutes offset from UTC

  UserProfile({
    required this.id,
    required this.name,
    required this.email,
    required this.joinedAt,
    this.dailyGoalMinutes = 10,
    this.dailyGoalAyahs = 0,
    this.level = 1,
    this.xp = 0,
    this.longestStreak = 0,
    this.avatarUrl,
    this.timezoneOffset = 0,
  });

  factory UserProfile.fromJson(Map<String, dynamic> json) =>
      _$UserProfileFromJson(json);

  Map<String, dynamic> toJson() => _$UserProfileToJson(this);
}

/// Single recitation session
@JsonSerializable()
class Session {
  final String id;
  final String userId;
  final int surah;
  final int startAyah;
  final int endAyah;
  final double durationMinutes;
  final String date; // UTC date (YYYY-MM-DD)
  final double accuracyScore; // 0-100
  final String mode; // "recitation" | "tajweed_lesson" | "review"
  final DateTime createdAt;
  final int xpEarned;

  Session({
    required this.id,
    required this.userId,
    required this.surah,
    required this.startAyah,
    required this.endAyah,
    required this.durationMinutes,
    required this.date,
    required this.accuracyScore,
    required this.mode,
    required this.createdAt,
    this.xpEarned = 0,
  });

  factory Session.fromJson(Map<String, dynamic> json) =>
      _$SessionFromJson(json);

  Map<String, dynamic> toJson() => _$SessionToJson(this);
}

/// Memorization progress for a surah
@JsonSerializable()
class MemorizationProgress {
  final String userId;
  final int surah;
  final int ayahCountMemorized;
  final DateTime? lastReviewedAt;
  final Map<int, int> highAccuracySessions; // {ayahNum: count}

  MemorizationProgress({
    required this.userId,
    required this.surah,
    this.ayahCountMemorized = 0,
    this.lastReviewedAt,
    this.highAccuracySessions = const {},
  });

  factory MemorizationProgress.fromJson(Map<String, dynamic> json) =>
      _$MemorizationProgressFromJson(json);

  Map<String, dynamic> toJson() => _$MemorizationProgressToJson(this);
}

/// Daily progress metrics
@JsonSerializable()
class DailyProgress {
  final String date; // YYYY-MM-DD
  final int totalMinutes;
  final int sessionsCount;
  final double accuracyAverage;
  final String status; // "not_started" | "in_progress" | "completed"

  DailyProgress({
    required this.date,
    this.totalMinutes = 0,
    this.sessionsCount = 0,
    this.accuracyAverage = 0.0,
    this.status = "not_started",
  });

  factory DailyProgress.fromJson(Map<String, dynamic> json) =>
      _$DailyProgressFromJson(json);

  Map<String, dynamic> toJson() => _$DailyProgressToJson(this);
}

/// Streak information
@JsonSerializable()
class StreakInfo {
  final int currentStreakDays;
  final int longestStreakDays;
  final String? lastSessionDate; // YYYY-MM-DD

  StreakInfo({
    this.currentStreakDays = 0,
    this.longestStreakDays = 0,
    this.lastSessionDate,
  });

  factory StreakInfo.fromJson(Map<String, dynamic> json) =>
      _$StreakInfoFromJson(json);

  Map<String, dynamic> toJson() => _$StreakInfoToJson(this);
}

/// Level and XP information
@JsonSerializable()
class LevelInfo {
  final int levelNumber;
  final int xpTotal;
  final int xpIntoLevel;
  final int xpForNextLevel;
  final double percentToNext; // 0-100

  LevelInfo({
    required this.levelNumber,
    required this.xpTotal,
    required this.xpIntoLevel,
    required this.xpForNextLevel,
    required this.percentToNext,
  });

  factory LevelInfo.fromJson(Map<String, dynamic> json) =>
      _$LevelInfoFromJson(json);

  Map<String, dynamic> toJson() => _$LevelInfoToJson(this);
}

/// Top memorized surah
@JsonSerializable()
class TopSurah {
  final int surahNumber;
  final String surahName;
  final double memorizedPercent;
  final int ayahCountMemorized;
  final int totalAyahs;

  TopSurah({
    required this.surahNumber,
    required this.surahName,
    required this.memorizedPercent,
    required this.ayahCountMemorized,
    required this.totalAyahs,
  });

  factory TopSurah.fromJson(Map<String, dynamic> json) =>
      _$TopSurahFromJson(json);

  Map<String, dynamic> toJson() => _$TopSurahToJson(this);
}

/// Aggregated metrics for home screen
@JsonSerializable()
class HomeMetrics {
  final bool isNewUser;
  final Map<String, dynamic> daily;
  final Map<String, dynamic> streak;
  final Map<String, dynamic> week;
  final Map<String, dynamic> level;
  final Map<String, dynamic> memorization;
  final Map<String, dynamic>? lastSession;

  HomeMetrics({
    required this.isNewUser,
    required this.daily,
    required this.streak,
    required this.week,
    required this.level,
    required this.memorization,
    this.lastSession,
  });

  factory HomeMetrics.fromJson(Map<String, dynamic> json) =>
      _$HomeMetricsFromJson(json);

  Map<String, dynamic> toJson() => _$HomeMetricsToJson(this);

  // Convenience getters
  int get dailyMinutes => (daily['minutes'] as num?)?.toInt() ?? 0;
  int get dailyGoal => (daily['goalMinutes'] as num?)?.toInt() ?? 10;
  double get dailyCompletion =>
      (daily['completionRatio'] as num?)?.toDouble() ?? 0.0;
  String get dailyStatus => (daily['status'] as String?) ?? "not_started";

  int get currentStreak => (streak['current'] as num?)?.toInt() ?? 0;
  int get longestStreak => (streak['longest'] as num?)?.toInt() ?? 0;

  int get weekMinutes => (week['totalMinutes'] as num?)?.toInt() ?? 0;
  int get weekDaysActive => (week['daysActive'] as num?)?.toInt() ?? 0;

  int get userLevel => (level['level'] as num?)?.toInt() ?? 1;
  int get totalXp => (level['xp'] as num?)?.toInt() ?? 0;
  double get percentToNextLevel =>
      (level['percentToNext'] as num?)?.toDouble() ?? 0.0;

  double get overallMemorization =>
      (memorization['overallPercent'] as num?)?.toDouble() ?? 0.0;

  List<Map<String, dynamic>> get topSurahs =>
      (memorization['topSurahs'] as List?)?.cast<Map<String, dynamic>>() ?? [];
}

File: backend/gamification_models.py
# -*- coding: utf-8 -*-
"""
Gamification Data Models for ReciteRight
Handles user profiles, sessions, memorization progress, streaks, levels, etc.
"""
from dataclasses import dataclass, field, asdict
from typing import List, Optional, Dict, Any
from datetime import datetime, timedelta
import json


@dataclass
class UserProfile:
    """User profile with gamification metadata"""
    id: str
    name: str
    email: str
    joined_at: str  # ISO 8601 UTC timestamp
    daily_goal_minutes: int = 10
    daily_goal_ayahs: int = 0
    level: int = 1
    xp: int = 0
    longest_streak: int = 0
    avatar_url: Optional[str] = None
    timezone_offset: int = 0  # Minutes offset from UTC

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'UserProfile':
        return cls(**data)


@dataclass
class Session:
    """Single recitation session"""
    id: str
    user_id: str
    surah: int
    start_ayah: int
    end_ayah: int
    duration_minutes: float
    date: str  # UTC date (YYYY-MM-DD)
    accuracy_score: float  # 0-100
    mode: str  # "recitation" | "tajweed_lesson" | "review"
    created_at: str  # ISO 8601 UTC timestamp
    xp_earned: int = 0

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Session':
        return cls(**data)


@dataclass
class MemorizationProgress:
    """Memorization progress for a surah"""
    user_id: str
    surah: int
    ayah_count_memorized: int = 0
    last_reviewed_at: Optional[str] = None  # ISO 8601 UTC timestamp
    high_accuracy_sessions: Dict[int, int] = field(default_factory=dict)  # {ayah_num: count}

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'MemorizationProgress':
        return cls(**data)


@dataclass
class DailyProgress:
    """Daily progress metrics"""
    date: str  # YYYY-MM-DD
    total_minutes: int = 0
    sessions_count: int = 0
    accuracy_average: float = 0.0
    status: str = "not_started"  # "not_started" | "in_progress" | "completed"

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


@dataclass
class WeeklySummary:
    """Weekly progress summary"""
    week_start: str  # YYYY-MM-DD (Monday)
    total_minutes: int = 0
    days_active: int = 0
    average_minutes_per_day: float = 0.0
    daily_breakdown: List[DailyProgress] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "week_start": self.week_start,
            "total_minutes": self.total_minutes,
            "days_active": self.days_active,
            "average_minutes_per_day": self.average_minutes_per_day,
            "daily_breakdown": [d.to_dict() for d in self.daily_breakdown]
        }


@dataclass
class StreakInfo:
    """Streak information"""
    current_streak_days: int = 0
    longest_streak_days: int = 0
    last_session_date: Optional[str] = None  # YYYY-MM-DD

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


@dataclass
class LevelInfo:
    """Level and XP information"""
    level_number: int
    xp_total: int
    xp_into_level: int
    xp_for_next_level: int
    percentage_to_next: float  # 0-100

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


@dataclass
class TopSurah:
    """Top memorized surah"""
    surah_number: int
    surah_name: str
    memorized_percent: float
    ayah_count_memorized: int
    total_ayahs: int

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


@dataclass
class HomeMetrics:
    """Aggregated metrics for home screen"""
    is_new_user: bool
    daily: Dict[str, Any]  # {minutes, goalMinutes, completionRatio, status}
    streak: Dict[str, Any]  # {current, longest}
    week: Dict[str, Any]  # {totalMinutes, daysActive, averagePerDay}
    level: Dict[str, Any]  # {xp, level, xpIntoLevel, xpForNextLevel, percentToNext}
    memorization: Dict[str, Any]  # {overallPercent, topSurahs}
    last_session: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        return {
            "isNewUser": self.is_new_user,
            "daily": self.daily,
            "streak": self.streak,
            "week": self.week,
            "level": self.level,
            "memorization": self.memorization,
            "lastSession": self.last_session,
        }

